dask/tests/test_base.py

```diff
@@ -405,6 +407,42 @@ def test_tokenize_ordered_dict():
     assert tokenize(a) != tokenize(c)
 
 
+ADataClass = dataclasses.make_dataclass("ADataClass", [("a", int)])
+BDataClass = dataclasses.make_dataclass("BDataClass", [("a", Union[int, float])])
+
+
+def test_tokenize_dataclass():
+    a1 = ADataClass(1)
+    a2 = ADataClass(2)
+    assert tokenize(a1) == tokenize(a1)
+    assert tokenize(a1) != tokenize(a2)
+
+    # Same field names and values, but dataclass types are different
+    b1 = BDataClass(1)
+    assert tokenize(a1) != tokenize(b1)
+
+    class SubA(ADataClass):
+        pass
+
+    assert dataclasses.is_dataclass(
+        SubA
+    ), "Python regression: SubA should be considered a dataclass"
+    assert tokenize(SubA(1)) == tokenize(SubA(1))
+    assert tokenize(SubA(1)) != tokenize(a1)
+
+    # Same name, same values, the definition has not changed: tokenize the same
+    ADataClassRedefined = dataclasses.make_dataclass("ADataClass", [("a", int)])
+    assert tokenize(a1) == tokenize(ADataClassRedefined(1))
+    assert tokenize(a2) != tokenize(ADataClassRedefined(1))
+
+    # Same name, same values, new definition: tokenize differently
+    ADataClassRedefinedDifferently = dataclasses.make_dataclass(
+        "ADataClass", [("a", Union[int, str])]
+    )
+    assert ADataClass is not ADataClassRedefinedDifferently
+    assert tokenize(a1) != tokenize(ADataClassRedefinedDifferently(1))
```

Comment:
@jcrist this test fails, because it turns out cloudpicke (and pickle) can't pickle dataclasses at protocol=0.

```
TypeError: cannot pickle '_DataclassParams' object: a class that defines __slots__ without defining __getstate__ cannot be pickled with protocol 0
```

So we fall back to just returning the name of the dataclass, which is exactly what I'd wanted to avoid (so if you redefine it in a notebook, the token changes).
https://github.com/dask/dask/blob/2f01fa4ec4af05700781f9c36723e791c17dfc9d/dask/base.py#L979-L990

Increasing the protocol level makes it pickleable. Though interestingly it also makes `assert tokenize(a1) == tokenize(ADataClassRedefined(1))` fail, which I don't mind too much. Inconsistent tokens for data that's actually the same just causes recompilation; token collisions for data that's actually different is what causes confusing errors.

Do you know why the protocol is pinned at 0?

Reply:
No idea, I think you should be fine to bump it up to protocol 4 (which is supported across all python versions we care about). We just want to make sure we pin it to something so that dask tokenizes the same across all python versions.

Reply:
Git blame turns up: https://github.com/dask/dask/pull/1118/commits/4474329bce9fbd9a3bd237eda704b418faaabc06 with no comments on why. I suspect @mrocklin picked protocol 0 to ensure compatibility across all python versions, even though at the time higher protocols did work on all versions we cared about. I wouldn't worry about it and just bump to protocol 4.

Reply:
@scharlottej13 has a similar PR where she arrived at the same conclusion https://github.com/dask/dask/pull/8527